pkg/diffs/diffs.go

```diff
@@ -0,0 +1,66 @@
+package diffs
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	prowconfig "k8s.io/test-infra/prow/config"
+)
+
+// Diffs holds the information where the differences can be found.
+type Diffs struct {
+	ConfigPath     string
+	JobConfigPath  string
+	DecorationPath string
+}
+
+// NewDiffs creates a new Diffs{}
+func NewDiffs(configPath, jobConfigPath, decorationPath string) *Diffs {
+	return &Diffs{
+		ConfigPath:     configPath,
+		JobConfigPath:  jobConfigPath,
+		DecorationPath: decorationPath,
+	}
+}
+
+// GetPresubmitsToExecute returns a mapping of repo to presubmits to execute.
+func (d *Diffs) GetPresubmitsToExecute() (map[string][]prowconfig.Presubmit, error) {
+	preSubmitsToExecuteMap := make(map[string][]prowconfig.Presubmit)
+
+	var prJobs map[string]map[string]prowconfig.Presubmit
+	var masterJobs map[string]map[string]prowconfig.Presubmit
+
+	prowMasterConfig, err := prowconfig.Load(d.ConfigPath, d.JobConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load master's Prow config: %v", err)
+	}
+
+	prowPRConfig, err := prowconfig.Load(fmt.Sprintf("%s/cluster/ci/config/prow/config.yaml", d.DecorationPath), fmt.Sprintf("%s/ci-operator/jobs/", d.DecorationPath))
+	if err != nil {
+		return nil, fmt.Errorf("failed to load PR's Prow config: %v", err)
+	}
+
+	masterJobs = getJobsByRepoAndName(prowMasterConfig.JobConfig.Presubmits)
+	prJobs = getJobsByRepoAndName(prowPRConfig.JobConfig.Presubmits)
+	for repo, jobMap := range prJobs {
+		for jobName, job := range jobMap {
+			if !equality.Semantic.DeepEqual(masterJobs[repo][jobName].Spec, job.Spec) {
```

Comment:
I think this will crash on newly added jobs. Newly added job's jobName won't have an appropriate master presubmit item, and we will attempt to access such item's Spec.

Also, are we really sure we do not want to rehearse jobs changed in non-Spec fields? Are all these fields that harmless?

@droslean 's comment in the original #50:
> We will have to be more specific in which fields we need to execute the job. Here I only check the .spec.

What do you mean about being more specific?

Reply:
Accessing non-existing keys yield an uninitialized map followed by an uninitialized `ProwJobSpec`, which should be fine if `DeepEqual` can handle the later.

It may be faster to explicitly check for their existence, but the tool seemed fast enough in my tests.

Reply:
Sorry, I spent too much time in my life using languages where uninitialized stuff bites you in the butt.

Reply:
There are dozens of fields outside of spec. Why not diff on them all? If I change an `agent: kubernetes` job to `agent: jenkins` I would expect this rehearsal to catch it.